### PR TITLE
Charge for reservation holding

### LINKED ISSUE #1830: Charge gas for reservation holding
### Problem

Currently there is no charging for storing gas reservation

### Steps

.

### Possible Solution

_No response_

### Notes

_No response_

### Relevant Log Output

<details><summary>Click to expand/collapse</summary>
<p>

```
<Paste your log here>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by ark0f**
Resolves #1830

### CODE REVIEW

File: core-processor/src/ext.rs [hunk 1]
```diff
@@ -613,9 +613,18 @@ impl EnvExt for Ext {
     fn reserve_gas(&mut self, amount: u64, duration: u32) -> Result<ReservationId, Self::Error> {
         self.charge_gas_runtime(RuntimeCosts::ReserveGas)?;
 
-        let common_charge = self.context.gas_counter.reduce(amount);
-        if common_charge == ChargeResult::NotEnough {
-            return Err(ExecutionError::InsufficientGasForReservation.into());
+        // TODO: amount != 0
```
Comment (breathx) [inline core-processor/src/ext.rs:616]: resolve your own todos

File: core-processor/src/ext.rs [hunk 2]
```diff
@@ -613,9 +613,18 @@ impl EnvExt for Ext {
...
+
+        if self.context.gas_counter.reduce(amount) == ChargeResult::NotEnough {
+            return self.return_and_store_err(Err(ExecutionError::InsufficientGasForReservation));
+        }
+
+        let reserve = u64::from(self.context.reserve_for.saturating_add(duration))
+            .saturating_mul(self.context.reservation);
+
+        if self.context.gas_counter.reduce(reserve) == ChargeResult::NotEnough {
+            return self.return_and_store_err(Err(ExecutionError::InsufficientGasForReservation));
```
Comment (breathx) [inline core-processor/src/ext.rs:627]: why not add them up to reduce once?

File: core/src/message/context.rs [hunk 3]
```diff
@@ -84,6 +84,8 @@ impl ContextSettings {
         self.waking_fee
     }
 
+    // TODO: introduce reservation fee
```
Comment (breathx) [inline core/src/message/context.rs:87]: todo

File: core/src/reservation.rs [hunk 4]
```diff
@@ -159,21 +148,38 @@ impl GasReserver {
     }
 
     /// Convert into gas reservation map.
-    pub fn into_map<F>(self, duration_into_expiration: F) -> GasReservationMap
+    pub fn into_map<F>(self, start: u32, duration_into_expiration: F) -> GasReservationMap
```
Comment (breathx) [inline core/src/reservation.rs:151]: I'd rename to not mess up with start from existing reservation state

File: pallets/gear/src/manager/journal.rs [hunk 5]
```diff
@@ -472,16 +472,13 @@ where
             unreachable!("Threshold for reservation invalidated")
```
Comment (breathx) [inline pallets/gear/src/manager/journal.rs:472]: without validation on zero duration you can run into this unreachable - need test also

File: pallets/gear/src/manager/mod.rs [hunk 6]
```diff
@@ -51,12 +51,15 @@ mod task;
 pub use journal::*;
```
Comment (breathx) [inline pallets/gear/src/manager/mod.rs:51]: https://github.com/gear-tech/gear/blob/23aa9ef2c84062056589d0b54250dce92b6cef1b/pallets/gear/src/manager/task.rs#L145-L147
This seems tricky for me, I'd go with renaming into `remove_gas_reservation_impl` for manager's not-trait fn cause it's not intuitive atm and MAY depend on compiler (which fn to call, probability of inf recursion)

File: pallets/gear/src/tests.rs [hunk 7]
```diff
@@ -6982,6 +6982,174 @@ fn system_reservation_zero_amount_panics() {
...
+        assert_ok!(Gear::send_message(
+            RuntimeOrigin::signed(USER_1),
+            pid,
+            HandleAction::Unreserve.encode(),
+            spent_gas,
+            0
+        ));
+
+        run_to_block(3, None);
+
```
Comment (breathx) [inline pallets/gear/src/tests.rs:7029]: assert balances here?

Comment (breathx) [reply pallets/gear/src/tests.rs:7029]: of accounts I mean, not reservations

Comment (breathx) [reply pallets/gear/src/tests.rs:7029]: and for other blocks if needed also

File: pallets/gear/src/tests.rs [hunk 8]
```diff
@@ -6982,6 +6982,174 @@ fn system_reservation_zero_amount_panics() {
...
+        // gas not yet unreserved automatically
+        let map = get_reservation_map(pid).unwrap();
+        assert_eq!(map.len(), 2);
+
+        run_to_block(2 + 3, None);
+
+        // gas unreserved automatically
+        let map = get_reservation_map(pid).unwrap();
+        assert_eq!(map.len(), 1);
```
Comment (breathx) [inline pallets/gear/src/tests.rs:7044]: balances

File: pallets/gear/src/tests.rs [hunk 9]
```diff
@@ -6982,6 +6982,174 @@ fn system_reservation_zero_amount_panics() {
...
+
+        // check task is exist yet
+        let (reservation_id, slot) = map.iter().next().unwrap();
+        let task = ScheduledTask::RemoveGasReservation(pid, *reservation_id);
+        assert!(TaskPoolOf::<Test>::contains(
+            &BlockNumberFor::<Test>::from(slot.finish),
+            &task
+        ));
+
+        // `gr_exit` occurs
```
Comment (breathx) [inline pallets/gear/src/tests.rs:7054]: just for confidence, do we have test for wait+wake in init and further termination for common reservations and system one?

Comment (breathx) [reply pallets/gear/src/tests.rs:7054]: and gr_exit test for system reservation with a)wait+wake and b)during the single execution

File: runtime/gear/src/lib.rs [hunk 10]
```diff
@@ -97,7 +97,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     impl_name: create_runtime_str!("gear"),
     apis: RUNTIME_API_VERSIONS,
     authoring_version: 1,
-    spec_version: 230,
+    spec_version: 240,
```
Comment (breathx) [inline runtime/gear/src/lib.rs:100]: spec_version: 100

**Review by breathx — CHANGES_REQUESTED** [review]

**Review by breathx — CHANGES_REQUESTED** [review]

File: pallets/gear/src/manager/journal.rs [hunk 11]
```diff
@@ -509,12 +511,15 @@ where
...
+            prog.gas_reservation_map = reserver.into_map(
+                Pallet::<T>::block_number().unique_saturated_into(),
+                |duration| {
+                    HoldBound::<T>::by(CostsPerBlockOf::<T>::reservation())
+                        .duration(BlockNumberFor::<T>::from(duration))
+                        .expected()
+                        .unique_saturated_into()
+                },
+            );
             common::set_program(pid, prog);
```
[the remark this answers is not in the document]
Comment (ark0f) [reply pallets/gear/src/manager/journal.rs:523]: And we don't do this

**Review by breathx — APPROVED** [review]

File: examples/binaries/signal-entry/src/lib.rs [hunk 12]
```diff
@@ -108,10 +111,28 @@ mod wasm {
...
             }
+            HandleAction::WaitAndExit => {
+                if DO_EXIT {
+                    msg::send_bytes(msg::source(), b"wait_and_exit", 0).unwrap();
+                    exec::exit(msg::source());
+                }
+
+                DO_EXIT = !DO_EXIT;
+
+                exec::system_reserve_gas(900).unwrap();
```
[the remark this answers is not in the document]
Comment (ark0f) [reply examples/binaries/signal-entry/src/lib.rs:122]: #1951

**Pull request merged**